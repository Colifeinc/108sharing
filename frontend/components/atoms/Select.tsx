import * as React from 'react'
import FormItem from 'components/layouts/FormItem'
import classNames from 'classnames'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome/'
import { faChevronDown } from '@fortawesome/free-solid-svg-icons/faChevronDown'

interface IOption {
  value: string
  label: string
}
interface IProps {
  name: string
  options: readonly IOption[]
  defaultValue: string
  label?: null | string
  error?: null | string
  required?: boolean
  selectRef?: any
  onChangeHandler?: (event) => void
  onBlurHandler?: (event) => void
}

const Select: React.FC<IProps> = ({
  name,
  options,
  defaultValue,
  label,
  error,
  required,
  selectRef,
  onChangeHandler,
  onBlurHandler,
}) => {
  return (
    <FormItem label={label} required={required} error={error}>
      <label className="relative block">
        <select
          className={classNames([
            'appearance-none m-0 bg-transparent outline-none text-current',
            'relative z-1 w-full py-1 px-3 leading-8 rounded-md border-2 border-transparent',
            'focus:shadow-none focus:border-2 focus:border-primary',
          ])}
          id={name}
          name={name}
          required={required}
          defaultValue={defaultValue}
          ref={selectRef || null}
          onChange={onChangeHandler}
          onBlur={onBlurHandler}
        >
          {options.map(option => (
            <option key={option.value} value={option.value}>
              {option.label}
            </option>
          ))}
        </select>
        {/* for focus style of select*/}
        <div className="absolute inset-0 rounded-md border border-neutral-300" />
        <div className="absolute right-0 inset-y-0 flex items-center mr-4">
          <FontAwesomeIcon className="text-sm" icon={faChevronDown} />
        </div>
      </label>
    </FormItem>
  )
}

export default Select
