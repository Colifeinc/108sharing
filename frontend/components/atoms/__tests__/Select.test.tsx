import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import * as React from 'react'
import Select from '../Select'

jest.mock('../../../core/i18n')

describe('atoms/Select.tsx', () => {
  const onChangeHandler = jest.fn()
  const onBlurHandler = jest.fn()
  const options = [
    { value: 'value1', label: 'option1' },
    { value: 'value2', label: 'option2' },
    { value: 'value3', label: 'option3' },
  ]
  const props = {
    name: 'Select Box',
    defaultValue: 'value2',
    label: 'Select',
    error: null,
    required: true,
    options,
    onChangeHandler,
    onBlurHandler,
  }

  it('should matches to the snapshot', () => {
    const wrapper = shallow(<Select {...props} />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('should has passed options', () => {
    const wrapper = shallow(<Select {...props} />)
    expect(wrapper.find('option[value="value1"]').length).toBe(1)
    expect(wrapper.find('option[value="value2"]').length).toBe(1)
    expect(wrapper.find('option[value="value3"]').length).toBe(1)
  })

  // it('should call handlers', () => {
  //   const wrapper = shallow(<Select {...props} />)
  //   wrapper.find('select').simulate('change')
  //   expect(onChangeHandler).toHaveBeenCalled()
  //   wrapper.find('select').simulate('blur')
  //   expect(onBlurHandler).toHaveBeenCalled()
  // })
})
