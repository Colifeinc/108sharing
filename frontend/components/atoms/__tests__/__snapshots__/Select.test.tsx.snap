// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`atoms/Select.tsx should matches to the snapshot 1`] = `
<FormItem
  error={null}
  label="Select"
  required={true}
>
  <label
    className="relative block"
  >
    <select
      className="appearance-none m-0 bg-transparent outline-none text-current relative z-1 w-full py-1 px-3 leading-8 rounded-md border-2 border-transparent focus:shadow-none focus:border-2 focus:border-primary"
      defaultValue="value2"
      id="Select Box"
      name="Select Box"
      onBlur={[MockFunction]}
      onChange={[MockFunction]}
      required={true}
    >
      <option
        key="value1"
        value="value1"
      >
        option1
      </option>
      <option
        key="value2"
        value="value2"
      >
        option2
      </option>
      <option
        key="value3"
        value="value3"
      >
        option3
      </option>
    </select>
    <div
      className="absolute inset-0 rounded-md border border-neutral-300"
    />
    <div
      className="absolute right-0 inset-y-0 flex items-center mr-4"
    >
      <FontAwesomeIcon
        border={false}
        className="text-sm"
        fixedWidth={false}
        flip={null}
        icon={
          Object {
            "icon": Array [
              448,
              512,
              Array [],
              "f078",
              "M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z",
            ],
            "iconName": "chevron-down",
            "prefix": "fas",
          }
        }
        inverse={false}
        listItem={false}
        mask={null}
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        swapOpacity={false}
        symbol={false}
        title=""
        transform={null}
      />
    </div>
  </label>
</FormItem>
`;
